operator/src/controller.rs

```diff
@@ -0,0 +1,228 @@
+///
+/// # Examples
+///
+/// ```no_run
+///     let (client, namespace): (Client, String) = deployment::try_default().await?;
+///     controller::run(client, &namespace).await;
+/// ```
+pub async fn run(client: Client, default_namespace: &str) {
+    let api: Api<H2O> = Api::all(client.clone());
+    Controller::new(api.clone(), ListParams::default())
+        .owns(api, ListParams::default())
+        .run(
+            reconcile,
+            error_policy,
+            Context::new(ContextData::new(client.clone(), default_namespace.to_string())),
+        )
+        .for_each(|res| async move {
+            match res {
+                Ok(o) => info!("Reconciled {:?}", o),
+                Err(e) => info!("Reconcile failed: {}", e),
+            };
+        })
+        .await;
+}
+
+/// Context data inserted into the reconciliation handler with each call.
+struct ContextData {
+    /// Kubernetes client to manipulate Kubernetes resources
+    client: Client,
+    /// Default namespace to deploy resources to - unless explicitly specified by the user
+    default_namespace: String,
+}
+
+impl ContextData {
+    /// Creates a new instance of `ContextData`.
+    ///
+    /// # Arguments
+    ///
+    /// - `client` - Kubernetes client to manipulate Kubernetes resources
+    /// - `default_namespace` - Default namespace to deploy resources to - unless explicitly specified by the user
+    pub fn new(client: Client, default_namespace: String) -> Self {
+        ContextData { client, default_namespace }
+    }
+}
+
+/// Action to be taken by the controller if there is a new event on `H2O` resource.
+enum ControllerAction {
+    /// Create a new H2O cluster
+    Create,
+    /// Delete resources of an existing H2O Cluster
+    Delete,
+    /// Updating existing H2O deployment is not supported - once H2O is clustered, it is immutable. Any events requiring on actions.
+    Noop,
+}
+
+/// Reconciliation logic router, called by the controller once per each event.
+/// See `ControllerAction` for details on routing logic.s
+///
+/// # Arguments
+/// `h2o` - The `H2O` resource, constructed automatically by the controller.
+/// `context` - Context to be injected into each reconciliation. Injected automatically by the controller.
+///
+/// # Examples
+/// No examples provided, as this method should be called only by the controller.
+async fn reconcile(h2o: H2O, context: Context<ContextData>) -> Result<ReconcilerAction, Error> {
+    match examine_h2o_for_actions(&h2o) {
+        ControllerAction::Create => {
+            create_h2o_deployment(&h2o, &context).await?;
+        }
+        ControllerAction::Delete => {
+            delete_h2o_deployment(&h2o, &context).await?;
+        }
+        ControllerAction::Noop => {
+            info!("No action taken for:\n{:?}", &h2o); // Log the whole incoming H2O description
+        }
+    }
+
+    return Ok(ReconcilerAction {
+        requeue_after: None,
```

Comment:
This is wrong. It does no harm directly, but no reconciliation means no periodical checks of running resources. 

- H2O pods may fail, as H2O might crash.
- When H2O fails, remaining resources should be cleaned and the H2O resource itself should be cleaned.
- H2O might not start properly - due to lack of resources available on nodes for example. 
- Headless service required for clustering or any other resource might NOT be initialized due to any possible fault on the Kubernetes cluster side. Such deployments should also be deleted.

Investigate possible random backoff using the [backoff](https://docs.rs/backoff/0.2.1/backoff/) crate.